### PR TITLE
Optimizer fast path

### PR DESCRIPTION
**Pull request by ggevay**
**Edit: I've done some benchmarking in release mode now, and the performance improvement in release mode is quite small (see [here](https://materializeinc.slack.com/archives/C0761MZ3QD9/p1722630312570519?thread_ts=1722544237.348389&cid=C0761MZ3QD9)), so I'm now not sure whether the complexity that this PR is introducing is actually worth it. We might want to hold off on merging this, so taking back to draft.**

**Edit 2: Now with @antiguru's other improvements, the speedup from this might be worth it. So marking ready for review.**

This PR is about running only a tiny subset of the optimizer for fast path queries. The approach is that when we are about to call `optimize_dataflow`, we quickly check whether we already have a simple enough plan that it would go to the fast path. If yes, then we call a special, tiny optimizer instead of the whole thing.

Commits:
1. Trivial comment fixes.
2. This is not strictly necessary, but makes the optimizer more robust. See commit msg.
3. The main point of the PR.
4. At the point of the peek optimization where I added the new call to `create_fast_path_plan`, we haven't yet eliminated `mz_now()`, and `create_fast_path_plan` might choke on that. This commit makes us ignore this error and just not take the fast path optimizer.

There is a tricky reason for the change in `src/environmentd/tests/testdata/http/ws`, I'm investigating the details now. It's probably not an issue, since the actual fast path plan that we arrive at is the same, just the non-fast-path plan is slightly different, but even there the physical plan would be the same, the diff is only in the MIR plan. Edit: Figured out exactly what's happening: The reason for the diff is that in the normal optimizer pipeline we call `ProjectionPushdown`, which puts the MFP in a non-canonical state (`Project` at the bottom), which is usually canonicalized by `CanonicalizeMfp`, but here `CanonicalizeMfp` is not able to put back the `Project` at the top for this query, because it calls `extract_non_errors_from_expr_mut`, which gives up when it sees the error in this MFP. But this doesn't matter because `create_fast_path_plan` then calls a different MFP extraction method: `extract_from_expression`, so the fast path plan will have the same MFP as before. (And `ProjectionPushdown` is not doing anything useful here, because MFP optimization will take care of the projections.)

### Motivation

https://materializeinc.slack.com/archives/C0761MZ3QD9/p1722545887422389?thread_ts=1722544237.348389&cid=C0761MZ3QD9

In my simplistic preliminary benchmarks (running a SELECT with a simple literal constraint on an indexed view locally in debug mode), I'm seeing a run time reduction from 20ms to 11ms. Edit: The difference was smaller in release mode. Edit 2: We should benchmark again, now that other parts of the system have been optimized in the meantime.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
  - [ ] Kicked off a limited Nightly: https://buildkite.com/materialize/nightly/builds/8931
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/environmentd/tests/testdata/http/ws [hunk 1]
```diff
@@ -409,7 +409,7 @@ ws-text
...
...                },\n                            {\n                              \"scalar_type\": \"Int32\",\n                              \"nullable\": false\n                            }\n                          ]\n                        }\n                      ],\n                      \"predicates\": [],\n                      \"projection\": [\n                        15\n                      ],\n                      \"input_arity\": 15\n                    }\n                  }\n                ]\n              }\n            }\n          ],\n          \"sources\": []\n        }\n      }\n    }\n  },\n  \"insights\": {\n    \"imports\": {\n      \"s697\": {\n        \"name\": {\n          \"schema\": \"mz_catalog\",\n          \"item\": \"mz_sources_ind\"\n        },\n        \"type\": \"compute\"\n      }\n    },\n    \"fast_path_clusters\": {},\n    \"fast_path_limit\": null,\n    \"persist_count\": []\n  },\n  \"cluster\": {\n    \"name\": \"mz_catalog_server\",\n    \"id\": {\n      \"System\": 2\n    }\n  },\n  \"redacted_sql\": \"SELECT '<REDACTED>' / '<REDACTED>' FROM [s456 AS mz_catalog.mz_sources] LIMIT '<REDACTED>'\"\n}","code":"MZ001","severity":"notice"}}
```
Comment (jkosh44) [inline src/environmentd/tests/testdata/http/ws:412]: Is this test useful for anyone? I have no idea what it was testing, what's changed in it, or if it's correct.

**Review by jkosh44 — APPROVED** [review]
Adapter code LGTM

File: src/environmentd/tests/testdata/http/ws (hunk 1 above)
Comment (ggevay) [reply src/environmentd/tests/testdata/http/ws:412]: See the last paragraph of the PR description.

**Comment by ggevay** [comment]
Bringing it out from draft, because, due to @antiguru's other improvements, the speedup from this might be worth it now.

Rebased on `main`.

I've split off the "Make LiteralConstraints specially handle empty possible_vals" commit into a different PR: https://github.com/MaterializeInc/materialize/pull/28748

File: src/transform/src/literal_constraints.rs
Comment (antiguru) [inline src/transform/src/literal_constraints.rs:1]: Should this be part of the literal constraints PR?

File: src/expr/src/relation.rs
Comment (antiguru) [inline src/expr/src/relation.rs:1]: Changes in this file seem unrelated/formatting.

File: src/adapter/src/optimize/peek.rs [hunk 2]
```diff
@@ -316,8 +314,30 @@ impl<'s> Optimize<LocalMirPlan<Resolved<'s>>> for Optimizer {
             &self.typecheck_ctx,
             &mut df_meta,
         );
+
+        // Let's already try creating a fast path plan. If successful, we don't need to run the
+        // whole optimizer pipeline, but just a tiny subset of it. (But we'll need to run
+        // `create_fast_path_plan` later again, because, e.g., running `LiteralConstraints` is still
+        // ahead of us.)
+        let use_fast_path_optimizer = match create_fast_path_plan(
```
Comment (antiguru) [inline src/adapter/src/optimize/peek.rs:322]: It might be a bit wasteful to create the fast path plan here and then run through the optimization pipeline again? Especially when we already here determine it's a constant or simple peek, we could probably use the result instead of running the optimizer at all? I'm fine with merging as-is, but we should take note.

**Review by antiguru — APPROVED** [review]
I think this looks great, and it has a positive performance impact. It reduces the average query latency. As expected, the tail latency stays largely unchanged.

File: src/expr/src/relation.rs
Comment (ggevay) [reply src/expr/src/relation.rs:1]: This is separated in the first commit ("Some trivial comment fixes") (which is also in the other PR). Is that ok, or would you prefer I open a separate PR for something like this?

Comment (antiguru) [reply src/expr/src/relation.rs:1]: Yes, please, it's distracting from the point of the PR.

Comment (ggevay) [reply src/expr/src/relation.rs:1]: Hmm. Then I'll remove this commit from this PR and open a separate PR just for that.

Comment (antiguru) [reply src/expr/src/relation.rs:1]: Thank you!

**Review by antiguru — APPROVED** [review]

File: src/adapter/src/optimize/peek.rs (hunk 2 above)
Comment (ggevay) [reply src/adapter/src/optimize/peek.rs:325]: > run through the optimization pipeline again

Do you mean "again", in the sense that we have already run the `optimize_mir_local` optimizer [above](https://github.com/MaterializeInc/materialize/blob/72a84836e8505131008d68353c9a8db882350a09/src/adapter/src/optimize/peek.rs#L186)? Running `optimize_dataflow` below still does meaningful things, i.e., not all the things that we want to happen for fast path peeks are done in `optimize_mir_local`. The most important thing that we still need to do is `LiteralConstraints`, which can turn a full index scan into an index lookup. (And then we need `CanonicalizeMfp` again, because `LiteralConstraints` can leave the MFP in a non-canonical form. And we also need `FoldConstants` again, because `LiteralConstraints` might have got us a constant, if it found contradicting literal constraints. And `ReduceScalars` is just good to run all the time for various obscure reasons, see the last paragraph of 3. [here](https://materializeinc.slack.com/archives/C0761MZ3QD9/p1720117362807379).)

(Note that `create_fast_path_plan` doesn't run any optimizations. It should be a cheaper things than even the `fast_path_optimizer`.)

A possible improvement that we could still do here is to not even run the `optimize_mir_local` for peeks that look like they are gonna be fast path, but just run `fast_path_optimizer` on the initial MIR plan. I tried to do this when originally working on this PR, and ran into some minor troubles, which I can't recall at the moment (maybe it was just that it would have required a somewhat larger refactoring of sequencing), so I just decided to keep this PR simple. But if the time comes when the optimizer run time becomes a significant portion of the run time of fast path peeks again, due to optimizations in other parts of the system, then I'm happy to revisit this. (I think I did some measurements with just hackily commenting out `optimize_mir_local`, and it showed that the run time of `optimize_mir_local` is significantly lower than the full optimizer pipeline, which this PR is eliminating.)

Comment (ggevay) [reply src/adapter/src/optimize/peek.rs:325]: I've created an issue for eliminating `optimize_mir_local` from fast path peeks, so that this doesn't get forgotten: https://github.com/MaterializeInc/materialize/issues/29501
